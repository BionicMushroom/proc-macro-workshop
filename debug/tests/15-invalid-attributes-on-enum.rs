use derive_debug::CustomDebug;

#[derive(CustomDebug)]
#[debug]
enum Test0 {}

#[derive(CustomDebug)]
#[debug = "0b{:08b}"]
enum Test1 {}

#[derive(CustomDebug)]
#[debug(bnd = "T: ::core::fmt::Debug")]
enum Test2<T> {
    Variant0(T),
}

#[derive(CustomDebug)]
#[debug(bound = "T: ::core::fmt::Debug")]
#[debug(bound = "T: ::core::fmt::Debug")]
enum Test3<T> {
    Variant0(T),
}

#[derive(CustomDebug)]
enum Test4 {
    Variant0(
        #[debug = "0b{:08b}"]
        #[debug = "0b{:08b}"]
        u8,
    ),
}

#[derive(CustomDebug)]
enum Test5<T> {
    Variant0(
        #[debug(bound = "T: ::core::fmt::Debug")]
        #[debug(bound = "T: ::core::fmt::Debug")]
        T,
    ),
}

#[derive(CustomDebug)]
enum Test6<T> {
    Variant0(
        #[debug(bound = "T: ::core::fmt::Debug")]
        #[debug = "0b{:08b}"]
        #[debug(bound = "T: ::core::fmt::Debug")]
        T,
    ),
}

#[derive(CustomDebug)]
#[debug(bound = "T: ::core::fmt::Debug")]
enum Test7<T> {
    Variant0(#[debug(bound = "T: ::core::fmt::Debug")] T),
}

#[derive(CustomDebug)]
enum Test8<T> {
    Variant0(#[debug(bound = "T: ::core::fmt::Debug")] T),
    Variant1(#[debug(bound = "T: ::core::fmt::Debug")] T),
}

#[derive(CustomDebug)]
enum Test9<T> {
    Variant0(
        #[debug(bound = "T: ::core::fmt::Debug")] T,
        #[debug(bound = "T: ::core::fmt::Debug")] T,
    ),
}

#[derive(CustomDebug)]
enum Test10<T> {
    Variant0(#[debug] T),
}

#[derive(CustomDebug)]
#[debug(bound = "T: ::core::fmt::Debug")]
enum Test11<T> {
    Variant0(#[debug] T),
}

#[derive(CustomDebug)]
enum Test12<T> {
    #[debug(bound = "T: ::core::fmt::Debug")]
    Variant0(T),
}

#[derive(CustomDebug)]
enum Test13<T> {
    #[debug = "0b{:08b}"]
    Variant0(T),
}

#[derive(CustomDebug)]
enum Test14<T> {
    #[debug]
    Variant0(T),
}

fn main() {}
