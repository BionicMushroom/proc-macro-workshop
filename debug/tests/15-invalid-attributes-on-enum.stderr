error: expected `debug(bound = "...")` bound attribute
 --> tests/15-invalid-attributes-on-enum.rs:4:3
  |
4 | #[debug]
  |   ^^^^^

error: `debug = "..."` format attribute is allowed only on struct and enum fields
 --> tests/15-invalid-attributes-on-enum.rs:8:1
  |
8 | #[debug = "0b{:08b}"]
  | ^^^^^^^^^^^^^^^^^^^^^

error: expected `debug(bound = "...")` bound attribute
  --> tests/15-invalid-attributes-on-enum.rs:12:3
   |
12 | #[debug(bnd = "T: ::core::fmt::Debug")]
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: multiple `debug(bound = "...")` bound attributes on the same enum
  --> tests/15-invalid-attributes-on-enum.rs:19:1
   |
19 | #[debug(bound = "T: ::core::fmt::Debug")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: multiple `debug = "..."` format attributes on the same enum field
  --> tests/15-invalid-attributes-on-enum.rs:28:9
   |
28 |         #[debug = "0b{:08b}"]
   |         ^^^^^^^^^^^^^^^^^^^^^

error: multiple `debug(bound = "...")` bound attributes on the same enum field
  --> tests/15-invalid-attributes-on-enum.rs:37:9
   |
37 |         #[debug(bound = "T: ::core::fmt::Debug")]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: multiple `debug(bound = "...")` bound attributes on the same enum field
  --> tests/15-invalid-attributes-on-enum.rs:47:9
   |
47 |         #[debug(bound = "T: ::core::fmt::Debug")]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `debug(bound = "...")` bound attribute is not allowed on enum fields if already specified on the enum itself
  --> tests/15-invalid-attributes-on-enum.rs:55:14
   |
55 |     Variant0(#[debug(bound = "T: ::core::fmt::Debug")] T),
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `debug(bound = "...")` bound attribute was already specified on this enum field type
  --> tests/15-invalid-attributes-on-enum.rs:61:14
   |
61 |     Variant1(#[debug(bound = "T: ::core::fmt::Debug")] T),
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `debug(bound = "...")` bound attribute was already specified on this enum field type
  --> tests/15-invalid-attributes-on-enum.rs:68:9
   |
68 |         #[debug(bound = "T: ::core::fmt::Debug")] T,
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: expected either `debug = "..."` format attribute or `debug(bound = "...")` bound attribute
  --> tests/15-invalid-attributes-on-enum.rs:74:16
   |
74 |     Variant0(#[debug] T),
   |                ^^^^^

error: expected `debug = "..."` format attribute
  --> tests/15-invalid-attributes-on-enum.rs:80:16
   |
80 |     Variant0(#[debug] T),
   |                ^^^^^

error: `debug(bound = "...")` bound attribute is not allowed on enum variants
  --> tests/15-invalid-attributes-on-enum.rs:85:5
   |
85 |     #[debug(bound = "T: ::core::fmt::Debug")]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `debug = "..."` format attribute is allowed only on struct and enum fields
  --> tests/15-invalid-attributes-on-enum.rs:91:5
   |
91 |     #[debug = "0b{:08b}"]
   |     ^^^^^^^^^^^^^^^^^^^^^

error: `debug = "..."` format attribute and `debug(bound = "...")` bound attribute are not allowed here
  --> tests/15-invalid-attributes-on-enum.rs:97:7
   |
97 |     #[debug]
   |       ^^^^^
